unit/util/expr_initializer.cpp

```diff
@@ -0,0 +1,274 @@
+    member.set_identifier("V" + std::to_string(i));
+    member.set_base_name("V" + std::to_string(i));
+    member.set_value(integer2bvrep(i, underlying_type.get_width()));
+    members.push_back(member);
+  }
+  return enum_type;
+}
+
+/// Create a type symbol, a type_tag and a symbol for the tag and populate with
+/// them the symbol_table in the environment.
+static symbolt create_tag_populate_env(
+  const typet &type,
+  expr_initializer_test_environmentt &env)
+{
+  const type_symbolt symbol{"my_type_symbol", type, ID_C};
+  symbolt tag_symbol = [&]() {
+    if(can_cast_type<c_enum_typet>(type))
+    {
+      const c_enum_tag_typet c_enum_tag_type{symbol.name};
+      return symbolt{"my_type_tag_symbol", c_enum_tag_type, ID_C};
+    }
+    else if(can_cast_type<struct_typet>(type))
+    {
+      const struct_tag_typet struct_tag_type{symbol.name};
+      return symbolt{"my_type_tag_symbol", struct_tag_type, ID_C};
+    }
+    else if(can_cast_type<union_typet>(type))
+    {
+      const union_tag_typet union_tag_type{symbol.name};
+      return symbolt{"my_type_tag_symbol", union_tag_type, ID_C};
+    }
+    UNREACHABLE;
+  }();
+  env.symbol_table.insert(symbol);
+  env.symbol_table.insert(tag_symbol);
+  return tag_symbol;
+}
+
+TEST_CASE("nondet_initializer boolean", "[core][util][expr_initializer]")
+{
+  auto test = expr_initializer_test_environmentt::make();
+  typet input = bool_typet{};
+  const auto result = nondet_initializer(input, test.loc, test.ns);
+  REQUIRE(result.has_value());
+  const auto expected = side_effect_expr_nondett{bool_typet{}, test.loc};
+  REQUIRE(result.value() == expected);
+}
+
+TEST_CASE("nondet_initializer signed_bv", "[core][util][expr_initializer]")
+{
+  auto test = expr_initializer_test_environmentt::make();
+  typet input = signedbv_typet{8};
+  const auto result = nondet_initializer(input, test.loc, test.ns);
+  REQUIRE(result.has_value());
+  const auto expected = side_effect_expr_nondett{signedbv_typet{8}, test.loc};
+  REQUIRE(result.value() == expected);
+}
+
+TEST_CASE("nondet_initializer c_enum", "[core][util][expr_initializer]")
+{
+  auto test = expr_initializer_test_environmentt::make();
+  const unsignedbv_typet enum_underlying_type{8};
+  const auto enum_type = make_c_enum_type(enum_underlying_type, 3);
+  const auto result = nondet_initializer(enum_type, test.loc, test.ns);
+  REQUIRE(result.has_value());
+  const auto expected = side_effect_expr_nondett{enum_type, test.loc};
+  REQUIRE(result.value() == expected);
+
+  // Repeat with the c_enum_tag_typet instead of the c_enum_typet it points to
+  const symbolt symbol = create_tag_populate_env(enum_type, test);
+  const auto tag_result = nondet_initializer(symbol.type, test.loc, test.ns);
+  REQUIRE(tag_result.has_value());
+  const auto expected_tag = side_effect_expr_nondett{symbol.type, test.loc};
+  REQUIRE(tag_result.value() == expected_tag);
+}
+
+TEST_CASE(
+  "nondet_initializer simple array_type",
```

Comment:
:pick: It might be nice to specify the type instead of calling it simple - "nondet_initializer on fixed-size array of signed 8 bit bytes."